### DISCUSSION TITLE
How can I allocate more ressources to Syslog-ng

### DISCUSSION THREAD

**Discussion by nicolasmf** [syslog-ng/syslog-ng#4295] [Q&A] [answered]
Hello,
I'm writing a configuration to aggregate logs received by UDP by devname, srcip, dstip and dstport fields. Here's my configuration : 

```
@version: 3.38
@include "scl.conf"

# Global Options
options {
        use_dns(no);
        dns-cache(no);
        keep-hostname(yes);
};

source s_test {
        udp( ip(<REDACTED>) port(5432) );
};

parser p_kv {
        kv-parser( prefix(".log.") pair-separator(" ") );
};

parser p_aggregate {
        grouping-by(
                key("${.log.devname}${.log.srcip}${.log.dstip}${.log.dstport}")
                aggregate(
                        value("MESSAGE" "${MESSAGE} log_count=$(- $(context-length) 1)")
                )
                timeout(10)
                inject-mode(aggregate-only) # Only send aggregated messages to the output
        );
};

destination d_test {
        file (
                "/DATA/messages.log"
                owner("root")
                group("root")
                perm(0777)
        );
};

log {
        source(s_test);
        parser(p_kv);
        parser(p_aggregate);
        destination(d_test);
};
```

As you can see on the tests I made (next section), the pass-throught configuration works well, but as I start using even only the parser(p_kv), some logs are dropped. I guess the problem comes from insuficient ressources for Syslog-ng. But when I look at the htop, I noticed that Syslog-ng doesn't use all ressources that are available in the machine.

Here is how syslog-ng is started : `ExecStart=/usr/sbin/syslog-ng -F --worker-threads=64` <= Maximum number of worker threads as mentionned in the [documentation](https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.16/administration-guide/81).

**How can I allocate more ressources to syslog-ng ?**

## Tests made
![image](https://user-images.githubusercontent.com/52258510/215497218-39cf036b-0a03-4103-b2eb-9153abdc0bd5.png)

## htop
![image](https://user-images.githubusercontent.com/52258510/215497847-29abbbdb-2f1d-489f-966e-054ab4f37af7.png)

Thank you in advance,

**Comment by bazsi** [comment, marked as answer]
Hi,

UDP is not a reliable protocol, therefore its receive buffer can overflow,
in which case the excess packets are dropped by the kernel.

You can monitor this using "netstat -anu" the UDP receive errors counters
show how many packets are dropped.

There are number of ways to mitigate this problem:

1) increase the size of the receive buffer. This would only work up to a
point, if the incoming rate is always higher than what syslog-ng is able to
consume the same will happen. The receive buffer size is under
/proc/sys/net/... Something like "rmem" size. I am unable to tell you the
exact name but it should be described in "man 7 UDP". Syslog-ng can even
increase this value by setting an option to the UDP source but there are
kernel limits that need to be increased.

2) you can use multiple UDP sources if you enable so-reuseaddr(), but this
only helps if you have multiple senders, as so-reuseaddr() load balances
incoming streams of data (a stream defined as source/destination addresses
and ports are the same)

3) ebpf can be used to improve ingestion rate but that means that you lose
ordering between messages. With reordering allowed you can get more
parallelism out of the processing of data.

4) in the recent weeks I was working to improve grouping-by() performance
wise. I didn't test how much faster it got, but if you can are willing to
test what I have, we can improve the situation.

As you can see all of these could help your alleviated, but using something
like tcp instead of UDP would be the best option, that could generate the
same pressure on the source side though.

--worker-threads argument does not really help here, as we are not loading
more CPUs with a single stream of messages.

Hope this helps,
Bazsi

**Reply by nicolasmf** [reply]
Hey Bazsi,
Thank you for your help. Unfortunately, going from UDP to TCP doesn't change anything, it seems like to be a lack of performance of the kv-parser and the aggregation because of the pass-throught config working properly.

I find strange that syslog-ng doesn't use a lot of cpu, maybe syslog-ng has some limits that I can extend ? 

Thank you,

**Comment by bazsi** [comment]
Can you give me the log sample you are sending to syslog-ng with the config
above? I'd like to look at what happens while processing it in more details.

**Reply by nicolasmf** [reply]
Sure !

The logs I'm working on are Fortinet FW logs.

Example log : 
`Feb  1 13:12:34 <Redacted_IP> <180>date=2023-01-23 time=13:57:22 devname="SOME_DEVICE_NAME" devid="..." logid="..." type="..." subtype="..." level="..." vd="..." eventtime=... tz="..." srcip=1.2.3.4 srcport=1234 srcintf="..." srcintfrole="..." dstip=1.2.3.4 dstport=1234 dstintf="..." dstintfrole="..." sessionid=... proto=... action="..." policyid=... policytype="..." poluuid="..." service="..." dstcountry="..." srccountry="..." trandisp="..." duration=... sentbyte=... rcvdbyte=... sentpkt=... rcvdpkt=... appcat="..." sentdelta=... rcvddelta=...`

If, in a 10 seconds interval, **8** logs have the same **devname**, **srcip**, **dstip** and **dstport**, this should be outputed : 

`Feb  1 13:12:34 <Redacted_IP> <180>date=2023-01-23 time=13:57:22 devname="SOME_DEVICE_NAME" devid="..." logid="..." type="..." subtype="..." level="..." vd="..." eventtime=... tz="..." srcip=1.2.3.4 srcport=1234 srcintf="..." srcintfrole="..." dstip=1.2.3.4 dstport=1234 dstintf="..." dstintfrole="..." sessionid=... proto=... action="..." policyid=... policytype="..." poluuid="..." service="..." dstcountry="..." srccountry="..." trandisp="..." duration=... sentbyte=... rcvdbyte=... sentpkt=... rcvdpkt=... appcat="..." sentdelta=... rcvddelta=... log_count=8`

**Comment by nicolasmf** [comment]
Hey,

Have you been able to check the process yet ?

Thank you for your help,

**Comment by bazsi** [comment]
Sorry I missed the message above. Let me look into this.

**Reply by nicolasmf** [reply]
No problem, thanks
